roadmap/implementers-guide/src/runtime/router.md

```diff
@@ -27,9 +110,108 @@ No initialization routine runs for this module.
 
 ## Routines
 
+The following routines are intended to be invoked by paras' upward messages.
+
+* `init_open_channel(recipient)`:
+  1. Check that the origin of this message is a para. We say that the origin of this message is the `sender`.
+  1. Check that the `sender` is not `recipient`.
+  1. Check that the `recipient` para exists.
+  1. Check that there is no existing open channel request from sender to recipient.
+  1. Check that the sum of the number of already opened HRMP channels by the `sender` (the size of the set found `HrmpEgressChannelsIndex` for `sender`) and the number of open requests by the `sender` (the value from `HrmpOpenChannelRequestCount` for `sender`) doesn't exceed the limit of channels (`config.hrmp_max_parachain_outbound_channels` or `config.hrmp_max_parathread_outbound_channels`) minus 1.
+  1. Reserve the deposit for the `sender` according to `config.hrmp_sender_deposit`
+  1. Add a new entry to `HrmpOpenChannelRequests` and increase `HrmpOpenChannelRequestCount` by 1 for the `sender`.
+      1. Set `sender_deposit` to `config.hrmp_sender_deposit`
+      1. Set `limit_used_places` to `config.hrmp_channel_max_places`
+      1. Set `limit_limit_used_bytes` to `config.hrmp_channel_max_size`
+
+* `accept_open_channel(i)`, `i` - is the index of open channel request:
+  1. Check that the designated open channel request exists
+  1. Check that the request's `recipient` corresponds to the origin of this message.
+  1. Reserve the deposit for the `recipient` according to `config.hrmp_recipient_deposit`
+  1. Set the request's `confirmed` flag to `true`.
+
+* `close_channel(sender, recipient)`:
+  1. Check that the channel between `sender` and `recipient` exists
+  1. Check that the origin of the message is either `sender` or `recipient`
+  1. Check that there is no existing intention to close the channel between `sender` and `recipient`.
```

Comment:
what exactly happens here? i could imagine a case where both `sender` and `recipient` initiate at the same time & race.

could it even count as a confirmed close?

Reply:
alternate behavior would be to ignore, and for paras to confirm close requests even when, from their perspective, they have already initiated a close.

Reply:
either way, would be nice to specify what happens here and what the corner cases are.

Reply:
There is no notion of close confirmation in this spec, since we decided to go ahead with unilateral closing with dropping messages atm to move things forward (Right?)

So there is no difference apart the value of `initiator` which I guess doesn't make any difference in the current spec and could be safely removed. In other words, if the race happens, even though the lost party's message fails to dispatch, the result is for all practical reasons is the same.